### ISSUE TITLE
Detect Kernel and OS version

### ISSUE THREAD

**Issue by marquiz** [kubernetes-sigs/node-feature-discovery#140]
These would help in placement of workloads which are dependent on certain OS or kernel version. 

I proposed implementing these in kubelet ([Issue #65850](https://github.com/kubernetes/kubernetes/issues/65850)), but, that request seems unlikely to be accepted. This feature could be implemented in NFD, instead.

Kernel version can be read from `/proc/sys/kernel/osrelease`. Kernel version would be split into multiple labels, in order to make it more practical to use as labelselectors need exact values, and, no comparison operators (i.e. "greater than" etc.) can be used. Suggested labels, with example values:
```
node.alpha.kubernetes-incubator.io/nfd-kernel-version=4.4.0-112-generic
node.alpha.kubernetes-incubator.io/nfd-kernel-version-x=4
node.alpha.kubernetes-incubator.io/nfd-kernel-version-xy=4.4
node.alpha.kubernetes-incubator.io/nfd-kernel-version-xyz=4.4.0
```

An initial RFD implementation of kernel version detection is available at:
https://github.com/marquiz/node-feature-discovery/tree/feature/kernel-version

OS version would be read from `/etc/os-release`. At least `ID` and `VERSION_ID` should be advertised, I think:
```
node.alpha.kubernetes-incubator.io/nfd-os-release-ID=ubuntu
node.alpha.kubernetes-incubator.io/nfd-os-release-VERSION_ID=16.04
```

An initial RFD implementation of OS version detection is available at:
https://github.com/marquiz/node-feature-discovery/tree/feature/os-release

**Comment by balajismaniam** [comment]
What is the use case for this feature?

**Comment by jeremyeder** [comment]
For example, whether or not the kernel has realtime capabilities.

**Comment by balajismaniam** [comment]
@jeremyeder I think that use case will be satisfied by #138?

**Comment by jeremyeder** [comment]
Interesting - ok. Probably either would work. For RH, we have 'rt' as a string in the kernel uname -r output, so for us it is a boolean, (either RT or not, one version or not).  I can see why #138 makes sense outside of my little ecosystem of kernels :-)

**Comment by marquiz** [comment]
> I think that use case will be satisfied by #138?

There are some cases I could think of:
* some ehancements/fixes that are not visible via kconfig flags
* user-specific custom kernel
* security

But, it is a fair question if the kconfig flags would be enough. I'm not necessarily super happy about these version number labels as they can cause problems because of the limitations of label selectors.

@lmdaly can you comment here?

**Comment by marquiz** [comment]
I created pull request #149 for the OS version part

**Comment by okartau** [comment]
Current name scheme is not intuitive: Seeing just nfd-kernel-version-x, or -xy does not tell easily what this means. One needs to see all 3 levels with value examples to get the point.
Also, now there is no isolated values from Y and Z fields.
What if I want to check for X == 4 and Y > 2 ?
I suggest, in addition to current values, also create isolated values with meaningful names, like
nfd-kernel-version-version: 4
nfd-kernel-version-major-rev: 4
nfd-kernel-version-minor-rev: 0

and yes, repeated "version" will look strange, but what form would be better?
As officially the first field is called "kernel version".

OTOH, this double set of labels may create some additional confusion.
But if we need to choose between schemed, I prefer isolated values, its more flexible to use and compare, IMHO.

**Comment by marquiz** [comment]
> Current name scheme is not intuitive: Seeing just nfd-kernel-version-x, or -xy does not tell easily what this means. One needs to see all 3 levels with value examples to get the point.

This is explained (with examples) in the README

> Also, now there is no isolated values from Y and Z fields.
> What if I want to check for X == 4 and Y > 2 ?
> I suggest, in addition to current values, also create isolated values with meaningful names, like
> nfd-kernel-version-version: 4
> nfd-kernel-version-major-rev: 4
> nfd-kernel-version-minor-rev: 0

ATM, I think, there is no point adding these *isolated* values as the k8s API does not support inequalities other than `!=` (like <, >) or other more complicated operators in selectors.

Those could be added later if the API started to support this use case

**Comment by marquiz** [comment]
I now created a PR implementing the kernel version part: #157

**Comment by marquiz** [comment]
> What is the use case for this feature? This will also cause UX issues like you mention @marquiz as label selectors don't support relational operators.

@balajismaniam: Would you reconsider? As stated above, nodeAffinity makes it possible to use `Lt` and `Gt` operators.

(should I edit the first message to reflect the latest design?)

**Comment by marquiz** [comment]
Just a minor thing, but, I started to think if it would make sense to use dot (`.`) character instead of dash (`-`) in the label name to separate version components?? I.e. something like
```
nfd-kernel-version.full=4.5.0-112-generic
nfd-kernel-version.major=4
nfd-kernel-version.minor=5
nfd-kernel-version.revision=0
```

Then similarly we would have kernel config labels like
```
nfd-kernel-config.NO_HZ
```

Any thoughts, or does it really matter?

**Comment by lmdaly** [comment]
So the the dot would be used for different components of a label? Would this impact other labels already introduced? I think the dot is a good differentiation between what is a label to a different component vs what is a feature of a component, but want to make sure its a well understood concept across all labels.

**Comment by marquiz** [comment]
Formulating it more formally, yes, it would be for separating different components of a feature. Currently, the label naming scheme (without the namespace part) is:
```
nfd-<FEATURE_SOURCE>-<FEATURE_NAME>
```

With this change we would refine the naming rule:
```
nfd-<FEATURE_SOURCE>-<FEATURE_NAME>.<FEATURE_COMPONENT>
```
(i don't know if *feature component* the most descriptive formulation,)

I don't think there are any existing labels that would violate this rule.

**Comment by marquiz** [comment]
I changed #149 and #157 to reflect this change.

Below is a list of currently advertised labels:
```
nfd-cpuid-<CPUID_FLAG>
nfd-iommy-enabled
nfd-memory-numa
nfd-network-sriov
nfd-network-sriov-configured
nfd-pstate-turbo
nfd-rdt-<RDT_FEATURE_NAME>
nfd-selinux-enabled
nfd-storage-nonrotationaldisk
```
Looking at it, the only one standing out is `nfd-network-sriov-configured`. But, I don't think that is would be a problem for the coherence.

**Comment by lmdaly** [comment]
Hmm, I'm wondering should it be sriov.enabled and sriov.configured?

**Comment by marquiz** [comment]
> Hmm, I'm wondering should it be sriov.enabled and sriov.configured?

I was thinking that one possibly slightly conflicting old label wouldn't be a problem.

But what you suggested would make more sense. And, now would be a good time to change that. Probably `sriov.present` and `sriov.configured` would be more accurate(?)

**Comment by lmdaly** [comment]
Yeah I think there should be a standard across all the labels, new and old so it would be good to align that now! Sounds good, or throwing in another option sriov.capable?

**Comment by marquiz** [comment]
I updated #149, the name of the feature source was changed from `os` to `system`. Labels names now being:
```
system-os_release-ID=ubuntu
system-os_release-VERSION_ID=16.04
```

**Comment by marquiz** [comment]
Closing this issue as os version detection was merged in #149
